Added basic VolumeManager support

GasimoCodes opened the pull request:
This PR adds experimental support for Unity's VolumeFramework system. This allows users to build their own effects or volume integrations within ToonRP.

TO-DO: Add volume support to built-in ToonRP Passes
TO-DO: Pass default values from PostProcessingPass to VolumeManager

GasimoCodes committed:
Added basic VolumeManager support

TO-DO: Load default values for PPPasses for blending
TO-DO: Add support for ToonRP Passes to support volumes

Delt06 commented on Assets/Samples/Volume Framework/Scripts/ToonSharpenTest.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace DELTation.ToonRP.PostProcessing.BuiltIn
+{
+    public class ToonSharpenTest : ToonPostProcessingPassBase
```
I suggest merging the new logic into the existing sharpen pass.

Delt06 replied:
Let's merge the copied sharpen pass/asset to the one in the package.
Also bring the corresponding volume component to the package folder.

Delt06 commented on Assets/Samples/Volume Framework/Scripts/ToonSharpenVolumeComponent.cs:
```diff
@@ -0,0 +1,18 @@
+[VolumeComponentMenu("ToonRP/Toon Sharpen")]
+public class ToonSharpenVolumeComponent : VolumeComponent
+{
+    public ClampedFloatParameter amount = new ClampedFloatParameter(0f, 0f, 10f);
```
`amount` -> `Amount`

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/ToonAdditionalCameraData.cs:
```diff
@@ -122,5 +133,72 @@ public void RestoreProjection()
+        /// <summary>
+        /// Returns true if this camera requires the volume framework to be updated every frame.
+        /// </summary>
+        public bool requiresVolumeFrameworkUpdate
```
`requiresVolumeFrameworkUpdate` -> `RequiresVolumeFrameworkUpdate`

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/ToonAdditionalCameraData.cs:
```diff
@@ -122,5 +133,72 @@ public void RestoreProjection()
+        /// Container for volume stacks in order to reuse stacks and avoid
+        /// creating new ones every time a new camera is instantiated.
+        /// </summary>
+        private static List<VolumeStack> s_CachedVolumeStacks;
```
`s_CachedVolumeStacks` -> `_cachedVolumeStacks`

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/ToonAdditionalCameraData.cs:
```diff
@@ -122,5 +133,72 @@ public void RestoreProjection()
+                m_VolumeStack = value;
+            }
+        }
+        VolumeStack m_VolumeStack = null;
```
`m_VolumeStack` -> `_volumeStack`

Delt06 replied:
Also the fields should all be at the top

GasimoCodes committed:
Fixed few naming conventions

GasimoCodes committed:
Added ToonRP VolumeManager default settings (Load from ToonPassAssets)

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/PostProcessing/ToonPostProcessingPassAsset.cs:
```diff
@@ -1,14 +1,16 @@
     public abstract class ToonPostProcessingPassAsset : ScriptableObject
     {
         public const string Path = "Toon RP/Post-Processing/";
+        const bool overrideSettings = true;
```
Please move this constant inside the `GetOrAddVolumeComponent` method

Delt06 commented on Assets/Samples/Volume Framework/Scripts/ToonSharpenVolumeComponent.cs:
```diff
@@ -0,0 +1,18 @@
+public class ToonSharpenVolumeComponent : VolumeComponent
+{
+    public ClampedFloatParameter Amount = new ClampedFloatParameter(0f, 0f, 10f);
+    public EnumParameter<PassOrder> Order = new EnumParameter<PassOrder>(PassOrder.PreUpscale);
```
I'm not sure whether the Order should be overridable, it may causes difficulties since changing it properly would require reordering passes.

GasimoCodes committed:
Integrated more effects to VolumeComponents

Converted:
Sharpen,
ToneMapping,
FilmGrain,
Bloom,
Vignette

GasimoCodes committed:
Added more Passes to VolumeFramework

Added:
- ScreenSpace Outlines
- Lookup Table
- Motion blur

Minor fixes for other stacks

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/PostProcessing/BuiltIn/ToonScreenSpaceOutlineComponent.cs:
```diff
@@ -0,0 +1,88 @@
+        /// </summary>
+        /// Since this is true it will not have any effect on the filter asset, it also allows easier casts
+        [HideInInspector]
+        public bool FilterEnabled = true;
```
It if does not affect anything, and the casts are done through field-by-field copying, do we really need it here?

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/PostProcessing/BuiltIn/ToonScreenSpaceOutlineComponent.cs:
```diff
@@ -0,0 +1,88 @@
+            : base(value, overrideState) { }
+
+        /*
+        public override void Interp(OutlineFilterReduced from, OutlineFilterReduced to, float t)
```
If it's not going to be used, better to remove commented code

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/PostProcessing/BuiltIn/ToonScreenSpaceOutlineComponent.cs:
```diff
@@ -0,0 +1,88 @@
+        public OutlineFilterParameter DepthFilter = new OutlineFilterParameter(new OutlineFilterReduced());
+        public OutlineFilterParameter NormalsFilter = new OutlineFilterParameter(new OutlineFilterReduced());
+        
+        // public BoolParameter boolParameter = new BoolParameter(false);
```
Remove commented code

GasimoCodes committed:
Fixed issues with applying Filter Enabled overrides, Cleaned up comments

GasimoCodes committed:
Added fog to volumes, moved GetSetVolumeComponent to Extension class

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/ToonCameraRenderer.cs:
```diff
@@ -555,6 +556,14 @@ private bool Cull(in ToonShadowSettings toonShadowSettings)
+            ToonFogComponent toonFog =  VolumeManager.instance.stack.GetComponent<ToonFogComponent>();
+
+            // Do RenderSettings related global overrides 
+            cmd.SetGlobalVector("unity_FogColor", toonFog.FogColor.value);
```
Use cached ID.

Delt06 commented on Packages/com.deltation.toon-rp/Runtime/PostProcessing/ToonBuiltinVolumeComponentContainer.cs:
```diff
@@ -0,0 +1,13 @@
+    /// <summary>
+    /// Holds cached volumes which dont belong to any specific PostPass or Extension
+    /// </summary>
+    public class ToonBuiltinVolumeComponentContainer
```
Seems like this is not used. If so, delete it.

GasimoCodes committed:
Removed ToonBuiltinVolumeComponentContainer, moved to unity_FogColor const

Delt06 reviewed (changes requested):
LGTM in general, just one small naming issue remaining.
